src/testplan/testplanTestRunner.ts

```diff
@@ -0,0 +1,225 @@
+import { ILogger } from '../logging/logger';
+import { IProcessExecution, runProcess } from '../processRunner'
+import { IDebugConfiguration, ITestRunner } from '../testRunner';
+import { empty } from '../utilities/collections';
+import { setDescriptionForEqualLabels } from '../utilities/tests';
+import { parseTestStates } from './testplanJunitTestStatesParser';
+import { parseTestSuites } from './testplanTestCollectionParser';
+
+// --- Testplan Exit Codes ---
+// 0: All tests were collected and passed successfully
+// 1: Tests were collected and run but some of the tests failed or none found
+const TESTPLAN_NON_ERROR_EXIT_CODES = [0, 1];
+
+const DISCOVERY_OUTPUT_PLUGIN_INFO = {
+    PACKAGE_PATH: path.resolve(__dirname, '../../resources/python'),
+    MODULE_NAME: 'vscode_python_test_adapter.testplan.discovery_output_plugin',
+};
+
+interface IRunArguments {
+    junitReportPath?: string;
+    argumentsToPass: string[];
+}
+
+export class TestplanTestRunner implements ITestRunner {
+
+    private readonly testExecutions: Map<string, IProcessExecution> = new Map<string, IProcessExecution>();
+
+    constructor(
+        public readonly adapterId: string,
+        private readonly logger: ILogger
+    ) { }
+
+    public cancel(): void {
+        this.testExecutions.forEach((execution, test) => {
+            this.logger.log('info', `Cancelling execution of ${test}`);
+            try {
+                execution.cancel();
+            } catch (error) {
+                this.logger.log('crit', `Cancelling execution of ${test} failed: ${error}`);
+            }
+        });
+    }
+
+    public async debugConfiguration(config: IWorkspaceConfiguration, test: string): Promise<IDebugConfiguration> {
+        const additionalEnvironment = await this.loadEnvironmentVariables(config);
+        const runArguments = this.getRunArguments(test, config.getTestplanConfiguration().testplanArguments);
+        return {
+            module: 'testplantest',
+            cwd: config.getCwd(),
+            args: runArguments.argumentsToPass,
+            env: additionalEnvironment,
+        };
+    }
+
+    public async load(config: IWorkspaceConfiguration): Promise<TestSuiteInfo | undefined> {
+        if (!config.getTestplanConfiguration().isTestplanEnabled) {
+            this.logger.log('info', 'TestPlan test discovery is disabled');
+            return undefined;
+        }
+        const additionalEnvironment = await this.loadEnvironmentVariables(config);
+        this.logger.log('info', `Discovering tests using python path '${config.pythonPath()}' in ${config.getCwd()}`);
+
+        const discoveryArguments = this.getDiscoveryArguments(config.getTestplanConfiguration().testplanArguments);
+        this.logger.log('info', `Running testplan with arguments: ${discoveryArguments.join(', ')}`);
+
+        const result = await this.runTestPlan(config, additionalEnvironment, discoveryArguments).complete();
+        const tests = parseTestSuites(result.output);
```

Comment:
In general, there are a lot of limitations to parsing stdout. For both unittest and pytest there are some wrappers/plugins that allow having a more or less consistent output. Are there any similar options for testplan?

We can leave this as is for now, but it's definitely worth looking into such things.

Reply:
testplan provides xml output, but only for test running, not for the discovery